### PR TITLE
Wrap duplicated-name definitions in anonymous namespace (except for iomgr)

### PR DESCRIPTION
**Pull request by vjpai**
This addresses some of the problems seen in #13832 . In particular, this wraps duplicated name definitions in anonymous namespaces except for those in `src/core/lib/iomgr` which is going through a separate overhaul anyway. Although these structs don't have external linkage using standard compile options, adding `-flto` to the gcc build flags to enable link-time optimization causes them to become subject to ODR rules.

Over time, more stuff from the files in `src/core` should be moved into anonymous namespaces anyway (in particular, to deprecate the use of `static` linkage which is non-preferred) so this is maybe the first step of a longer-term cleanup.

### CODE REVIEW

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE              FILE SIZE
 ++++++++++++++ GROWIN ++++++++++++++
  [ = ]       0 [None]    +424  +0.0%

  [ = ]       0 TOTAL     +424  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

File: src/core/lib/surface/completion_queue.cc [hunk 1]
```diff
@@ -780,10 +780,11 @@ typedef struct {
   bool first_loop;
 } cq_is_finished_arg;
 
-class ExecCtxNext : public grpc_core::ExecCtx {
+class ExecCtxNext final : public grpc_core::ExecCtx {
```
Comment (dgquintas) [inline src/core/lib/surface/completion_queue.cc]: why does this class (and `ExecCtxPluck`) need be final? Do these changes still belong to this PR?

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

File: src/core/lib/surface/completion_queue.cc (hunk 1 above)
Comment (vjpai) [reply src/core/lib/surface/completion_queue.cc]: No, we can take out the `final` part for now. That should be a different PR.

Comment (vjpai) [reply src/core/lib/surface/completion_queue.cc]: I removed the changes that were present in `completion_queue.cc` as they weren't needed for this PR.

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE              FILE SIZE
 ++++++++++++++ GROWIN ++++++++++++++
  [ = ]       0 [None]    +432  +0.0%

  [ = ]       0 TOTAL     +432  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Pull request merged**